library/include/Core.h

```diff
@@ -353,26 +354,33 @@ namespace DFHack
         friend struct CoreSuspendReleaseMain;
     };
 
-    class CoreSuspenderBase  : protected std::unique_lock<std::recursive_mutex> {
+    class CoreSuspenderBase : protected std::unique_lock< decltype(Core::CoreSuspendMutex) > {
     protected:
-        using parent_t = std::unique_lock<std::recursive_mutex>;
+        size_t lock_count = 0;
```

Comment:
The `CoreSuspender` in its current state doesn't have a need for keeping a lock count. `unique_lock` will throw an exception if locked more than once, disallowing the CoreSuspender being locked multiple times. In the destructor, unlock can be called if `unique_lock::owns_lock` as it tracks whether that particular instance owns the lock.

Reply:
It's a recursive mutex though -- it permits multiple locks by definition

Reply:
The recursive mutex does support multiple locks, but the unique_lock provides a wrapper around that mutex. The unique_lock prevents ever obtaining the lock twice by throwing an exception if a second attempt is made. Multiple CoreSuspender instances can all have locks at the same time thanks to the recursive mutex, but no single CoreSuspender instance can have multiple locks.

Reply:
It's a logic error for a `CoreSuspender` to get locked twice. Normally `CoreSuspender`s are locked at construction time, but we provide a deferred-lock variant that has be manually locked, so the `lock` method is public for that specific purpose.

There may be better ways to do this, and I'm totally open to suggestions for improvement.

I'll probably revise this to remove the lock counting; @NicksWorld has convinced me that it's not necessary

I'm thinking about removing the deferred-locking variant of `CoreSuspender` and creating a `ConditionalCoreSuspender` that does a `try_lock` at construction, and if it succeeds returns an object that holds a `CoreSuspender` which will be released when the `ConditionalCoreSuspender` is destroyed, and which evaluates to true if cast to boolean, and if it fails returns an object holding nothing and which evaluates to false if cast to boolean. This would allow something like `if (ConditionalCoreSuspender ccs) { ... }` for things that should not block if a core suspender cannot be acquired, while still providing RAII-like behavior and insuring that the lock is locked at most once and is unlocked when it goes out of scope

Reply:
i should also add that it's _not_ a logic error for a thread to acquire a `CoreSuspender` when it already has one, as long as it's a _different_ `CoreSuspender`. this will result in the underlying mutex being locked twice, which is harmless because the underlying mutex is recursive

i'm increasingly of the opinion that we should make the `lock` and `unlock` methods on `CoreSuspender` private and only allow `CoreSuspender`s to be locked at creation and unlocked at destruction, instead of relying on consumers to properly manage `CoreSuspender` state themselves, and move the cases that currently require the consumer to manage `CoreSuspender` state themselves to classes that abstract that management for them